### PR TITLE
Lecture 1: Setup & ProductDAO

### PR DESCRIPTION
**Pull request by antondolgopolik**
Add lecture 1 tasks

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,25 +1,78 @@
...
+        Product key = new Product();
+        key.setId(id);
+        synchronized (mutex) {
+            // Find index
+            int index = Collections.binarySearch(products, key, Comparator.comparing(Product::getId));
+            // Check if item is found
+            if (index >= 0) {
+                products.remove(index);
+            } else {
+                throw new NoProductWithSuchIdException(id);
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can we use getProduct(id) here instead of binarySearch?

Or even something like this:

products.removeIf(product -> id.equals(product.getId()));

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 2]
```diff
@@ -16,6 +19,36 @@ public void setup() {
...
+        Product product1 = new Product("code", "des", BigDecimal.ZERO, usd, 0, "url");
+        productDao.save(product1);
+    }
+
+    @Test
+    public void testGetProduct() throws NoProductWithSuchIdException {
+        productDao.getProduct(0L);
+        productDao.getProduct(10L);
+        productDao.getProduct(12L);
+    }
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: Do these methods without asserts really test anything?
Please add more tests for corner cases: save the product with the same values as existing product, etc.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 3]
```diff
@@ -1,25 +1,78 @@
...
+
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private final List<Product> products = new ArrayList<>();
+    private volatile long maxId;
+
+    private final Object mutex = new Object();
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can we simply use synchronized methods instead of mutex object?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 4]
```diff
@@ -1,25 +1,78 @@
...
+                .collect(Collectors.toList());
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        synchronized (mutex) {
+            product.setId(maxId++);
+            products.add(product);
+        }
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: What will happen if we add the product that already exists in the list?
A better approach may be to check for product.getId(), if it's null then add new product. If id is not null, then rewrite the existing product with the same id;

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 1 above)
Comment (antondolgopolik) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: We couldn't use getProduct as we needed list index of product with a specified ID to remove this product by that index.
We could use removeIf but with binary search it should work a little bit faster.
Now see new implementation with Map instead of List.

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java (hunk 2 above)
Comment (antondolgopolik) [reply src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: These tests only check if exceptions are thrown.
It's done.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 3 above)
Comment (antondolgopolik) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: I've replaced some synchronized statements by synchronized methods. Also I've removed mutex object, now synchronized statements use products.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 4 above)
Comment (antondolgopolik) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Now there is ability to save products with specified ID. If a product has the same content but different ID, then it's saved as another product.

**Review by SergeiStarikov — APPROVED** [review]
